samples/bluetooth/encrypted_advertising/central/src/central_ead.c

```diff
@@ -0,0 +1,534 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdint.h>
+#include <string.h>
+
+#include <zephyr/kernel.h>
+
+#include <zephyr/net/buf.h>
+#include <zephyr/logging/log.h>
+
+#include <zephyr/bluetooth/ead.h>
+#include <zephyr/bluetooth/att.h>
+#include <zephyr/bluetooth/gatt.h>
+#include <zephyr/bluetooth/addr.h>
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/bluetooth/uuid.h>
+#include <zephyr/bluetooth/bluetooth.h>
+
+#include "common/bt_str.h"
+
+#include "common.h"
+
+LOG_MODULE_REGISTER(ead_central_sample, CONFIG_BT_EAD_LOG_LEVEL);
+
+static uint8_t *received_data;
+static struct key_material keymat;
+
+static bt_addr_le_t peer_addr;
+static struct bt_conn *default_conn;
+
+static struct bt_conn_cb central_cb;
+static struct bt_conn_auth_cb central_auth_cb;
+
+static struct k_poll_signal conn_signal;
+static struct k_poll_signal paired_signal;
+static struct k_poll_signal sec_update_signal;
+static struct k_poll_signal device_found_cb_completed;
+
+/* GATT Discover data */
+static uint8_t gatt_disc_err;
+static uint16_t gatt_disc_end_handle;
+static uint16_t gatt_disc_start_handle;
+static struct k_poll_signal gatt_disc_signal;
+
+/* GATT Read data */
+static uint8_t gatt_read_err;
+static uint8_t *gatt_read_res;
+static uint16_t gatt_read_len;
+static uint16_t gatt_read_handle;
+static struct k_poll_signal gatt_read_signal;
+
+static bool data_parse_cb(struct bt_data *data, void *user_data)
+{
+	if (data->type == BT_DATA_ENCRYPTED_AD_DATA) {
+		int err;
+		struct net_buf_simple decrypted_buf;
+		size_t decrypted_data_size = BT_EAD_DECRYPTED_PAYLOAD_SIZE(data->data_len);
+		uint8_t decrypted_data[decrypted_data_size];
+
+		net_buf_simple_init_with_data(&decrypted_buf, &decrypted_data, decrypted_data_size);
+
+		err = bt_ead_decrypt(keymat.session_key, keymat.iv, data->data, data->data_len,
+				     decrypted_buf.data);
+		if (err < 0) {
+			LOG_ERR("Error during decryption.");
+		}
+
+		bt_data_parse(&decrypted_buf, &data_parse_cb, user_data);
```

Comment:
Nice idea to reuse the function here. 

Made me think: It is spec-compliant if someone encrypts some advertising data multiple times?

Reply:
It is. There is nothing that say you're not allowed to do it at least.